src/models/syncRule.ts

```diff
@@ -59,8 +75,14 @@ export default class SyncRule {
         );
     }
 
-    public updateTargetInstances(instances: string[]): void {
-        this.syncRule.builder.targetInstances = instances;
+    public updateTargetInstances(targetInstances: string[]): SyncRule {
+        return SyncRule.build({
+            ...this.syncRule,
+            builder: {
+                ...this.syncRule.builder,
+                targetInstances,
+            },
+        });
```

Comment:
Cool! I know at the beginning it feels cumbersome, but in my experience it pays off, on the long run, to have fully immutable models. By the way, this kind of boilerplate for nested updates can be simplified using lodash.fp, _[.set](https://gist.github.com/jfmengels/6b973b69c491375117dc#_setpath-value-object), but please don't change it, the method is not fully typed and we lose safety. In the function world, to work with nested structures, there is a concept called "lenses", for TS this is the best I know: https://github.com/utatti/lens.ts. Again, don't change it, just as a note, it's perfectly fine as it is, with just a level of nesting.

Reply:
WIll take a look on this lenses concept!